ls1intum/Artemis
# [Feature] Fetch latest result from server after submission timed-out

### Checklist

### Motivation and Context


Testers noticed recently that the new result for a programming exercise does not arrive sometimes. The reason for this is still unknown. In this PR we will add a fallback mechanism, which requests the latest result from the server if the result does not arrive in time (~2 minutes) over websocket.

### Description


- Add option to request _latest-result-with-feedbacks_ to also return the submission of the result
- Add fallback mechanism, which requests the latest result from the server if the latest submission times out. 
- Check that the fetched result belongs to the latest submission
- Notify all result subscribers with the fetched result if the submissions match

### Steps for Testing

We need to use the developer console to test this feature. I will illustrate the setup using Chrome in the following steps.
Read the instructions carefully before you start testing. Timing is crucial here.

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise (activate online code editor)
4. For a later testing step, open the Bamboo project, which contains the build plans for this exercise (e.g. by clicking on the link BASE or SOLUTION for the created exercise)
5. Start a participation for the exercise and open the online code editor
6. In the other tab open the build plan for this participation
7. Switch back to the code editor and open the developer console by pressing F12. Open the network tab
8. Solve one test case and submit. After the result is shown, wait two minutes and make sure that the request _latest-result-with-feedbacks?withSubmission=true_ does not show up in the network tab
9. Open the Network request blocking tool as seen in the screenshot below
![image](https://user-images.githubusercontent.com/16407766/107103222-1de7c180-681d-11eb-9eeb-bf1724aa1cff.png)
10. Add the following blocking rule but do not activate the request blocking yet
![image](https://user-images.githubusercontent.com/16407766/107103579-78cde880-681e-11eb-9de2-48daacefeed4.png)
11. Solve another test case and submit. Wait until _Building and testing_ is shown. Then activate the request blocking and refresh the browser page
--> After approximately two minutes the request _latest-result-with-feedbacks?withSubmission=true_ should show up in the network tab and the result with 2 solved test cases should be displayed.
12. Deactivate the request blocking and refresh the browser page
13. Remove the solution for one test case and submit
14. Stop the build plan for the corresponding participation in the other opened tab as seen below (might need to refresh to see the new running build)
![image](https://user-images.githubusercontent.com/16407766/107104051-605ecd80-6820-11eb-9346-bca4e8fe8c06.png)
--> After approximately two minutes check that the request _latest-result-with-feedbacks?withSubmission=true_ was executed. The result with two solved test cases should still be shown and the red re-trigger build button should be shown as seen below
![image](https://user-images.githubusercontent.com/16407766/107104538-6f467f80-6822-11eb-8b91-b24a8e94a4ec.png)
15. Test that the normal case, where the result arrives as usual still works and does not trigger the fallback after ~2 minutes
 
### Test Coverage



Add two test cases for the latest-result-with-feedbacks request. Coverage should remain approximately the same.
Added a client test case which tests the submission states after a successful fallback execution. Added assertions for another test case for the case, when the fallback mechanism does not fetch the right result for the pending submission
